### PR TITLE
Migrate from skills to interfaces

### PR DESCRIPTION
**Pull request by sergiusens**
This adds some sort of transparent migration logic from "skills" related
keywords to the ones belonging to "interfaces".

LP: #1549427

Signed-off-by: Sergio Schvezov <sergio.schvezov@ubuntu.com>

### CODE REVIEW

**Comment by sergiusens** [comment]
meh, cov is 93%... not sure what went wrong here

**Comment by sergiusens** [comment]
**this cannot land before we have an image**

File: snapcraft/tests/test_yaml.py [hunk 1]
```diff
@@ -63,6 +63,49 @@ def test_config_loads_plugins(self, mock_get_part, mock_loadPlugin):
         self.assertFalse(mock_get_part.called)
 
     @unittest.mock.patch('snapcraft.yaml.Config.load_plugin')
+    def test_config_works_with_skills(self, mock_loadPlugin):
```
Comment (zyga) [inline snapcraft/tests/test_yaml.py]: _with_interfaces

File: snapcraft/tests/test_yaml.py [hunk 2]
```diff
@@ -63,6 +63,49 @@ def test_config_loads_plugins(self, mock_get_part, mock_loadPlugin):
...
+
+parts:
+  part1:
+    plugin: go
+    stage-packages: [fswebcam]
+
+apps:
+   app1:
+     command: runme
+     uses:
```
Comment (zyga) [inline snapcraft/tests/test_yaml.py]: Is this intentional?

File: snapcraft/tests/test_yaml.py [hunk 3]
```diff
@@ -63,6 +63,49 @@ def test_config_loads_plugins(self, mock_get_part, mock_loadPlugin):
...
+            {'apps': {'app1': {'command': 'runme', 'slots': ['migration']}},
+             'architectures': [snapcraft.common.get_arch()],
+             'description': 'test',
+             'name': 'test',
+             'parts': {
+                'part1': {
+                    'snap': [], 'stage': [],
+                    'stage-packages': ['fswebcam']}},
+             'slots': {
+                'migration': {
```
Comment (zyga) [inline snapcraft/tests/test_yaml.py]: Ah, ignore me :)

File: snapcraft/yaml.py [hunk 4]
```diff
@@ -113,6 +113,10 @@ def __init__(self):
         self.after_requests = {}
 
         self.data = _snapcraft_yaml_load()
+
+        # To make the transition less painful
+        self._remap_skills()
```
Comment (zyga) [inline snapcraft/yaml.py]: Nitpicking now, perhaps ``_translate_skills_to_intrerfaces()``?

File: snapcraft/yaml.py [hunk 5]
```diff
@@ -152,6 +156,22 @@ def __init__(self):
         if 'architectures' not in self.data:
             self.data['architectures'] = [common.get_arch(), ]
 
+    def _remap_skills(self):
+        if 'uses' in self.data:
+            logger.warning(
+                "DEPRECATED: Instances of 'uses' remapped to 'slots'")
```
Comment (zyga) [inline snapcraft/yaml.py]: I'd display a link to the mailing list where Gustavo announces this in case anyone wonders what that is.

**Comment by zyga** [comment]
+1, with two optional notes

File: snapcraft/tests/test_yaml.py (hunk 2 above)
Comment (sergiusens) [inline snapcraft/tests/test_yaml.py]: yes, that is why the test is called *with skills*

you don't have to face angry users, I do :wink:

File: snapcraft/yaml.py (hunk 4 above)
Comment (sergiusens) [inline snapcraft/yaml.py]: I took the middle ground, `_remap_skills_to_interfaces`

File: snapcraft/yaml.py (hunk 5 above)
Comment (sergiusens) [inline snapcraft/yaml.py]: I'd rather wait for the documentation to surface on developer.ubuntu.com

File: snapcraft/tests/test_yaml.py (hunk 2 above)
Comment (kyrofa) [inline snapcraft/tests/test_yaml.py]: Yeah, good call.

File: snapcraft/yaml.py [hunk 6]
```diff
@@ -152,6 +156,22 @@ def __init__(self):
         if 'architectures' not in self.data:
             self.data['architectures'] = [common.get_arch(), ]
 
+    def _remap_skills_to_interfaces(self):
+        if 'uses' in self.data:
+            logger.warning(
+                "DEPRECATED: Instances of 'uses' remapped to 'slots'")
```
Comment (kyrofa) [inline snapcraft/yaml.py]: Ah, excellent.

**Comment by kyrofa** [comment]
:+1: from me

**Pull request merged**